README.md
![Graphene](docs/logo/GrapheneSignature.png)  
[![ktlint](https://img.shields.io/badge/code%20style-%E2%9D%A4-FF4081.svg)](https://ktlint.github.io/)  [![CircleCI](https://circleci.com/gh/graphene-monitoring/graphene.svg?style=svg)](https://circleci.com/gh/graphene-monitoring/graphene)   [![Join the chat at https://gitter.im/graphene-community/community](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/graphene-community/community)

## What is Graphene?
Graphene is a java version of [Graphite](https://graphiteapp.org). To
store metrics, it uses [Cassandra](https://github.com/apache/cassandra)
and [Elasticsearch](https://github.com/elastic/elasticsearch) instead of
[Whisper](https://github.com/graphite-project/whisper) to handle data in
a scalable way. It is designed to be compatible with Graphite's API and
easily integrates with Grafana's Graphite datasource.
In a cloud environment, the metric key increases exponentially. To cope
with this, each metric needs to be managed by time and removed or
archived according to the policy. Graphene is a system that supports
these methods later to help you manage your metrics cost-effectively. It
also supports multi-tenancy for metrics later, so that large numbers of
metrics can be maintained and managed in isolated clusters.

## Architecture Overview

![Graphene](docs/main/GrapheneArchitecture.png)

## API Support

### Graphite

|           API             | Support |
|---------------------------|---------|
| /render                   |    O    |
| /find                     |    O    |
| /metrics/find             |    O    |
| /tags/autoComplete/tags   |    O    |
| /tags/autoComplete/values |    O    |

### InfluxDB

|     API       | Support |
|---------------|---------|
| /write        |    O    |

## Prerequisite

| Storage       | Version |
|---------------|---------|
| Elasticsearch | 6.8.x   |
| Cassandra     | 3.11.x  |

## Commit convention
This project uses the git conventional commit rule provided by [conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/)

## Configuration
Please check [this document](https://github.com/graphene-monitoring/graphene/wiki/Configuration) for the Graphene configuration

## Thanks

Thanks, this project is useless without their work on **cyanite**, **graphite-api**, **graphite-cyanite**, **disthene-reader**, **disthene**

- Pierre-Yves Ritschard ([https://github.com/pyr](https://github.com/pyr))
- Bruno Renié ([https://github.com/brutasse](https://github.com/brutasse))
- EinsamHauer ([https://github.com/EinsamHauer/disthene-reader](https://github.com/EinsamHauer/disthene-reader))

Many people helped with the Graphene project.

- Keyn : Keyn had provided a lot of insights to improve Graphene system. Without Keyn, this project could not grow.
- Tiger : Tiger showed how to modify the open source and apply it to the production environment in order to improve the Graphite system operated by one server. This led to an understanding of the metric system and the ability to continue to expand the current system.
- El : El provided continuous insight based on superior technology than others. Based on this, the system can be viewed from various points of view instead of just one part.
- drunkencoding / panda84 : They developed and maintained the entire metric core system like Elasticsearch indexing logic and time life cycle in the metric key (path). Without them, the metric system would only be able to handle simple traffic.
- jerome89 ([https://github.com/jerome89]) : He found the grammar of Graphite functions in the disthene project. This helped us to create a better metric system. Additionally, he helped develop the faulty function of the graphene itself and the overall system development.

## License

The MIT License (MIT)

Copyright (C) 2015 Andrei Ivanov

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


RELEASE_NOTES.md
# 1.9.1

## Feature
- Improved tag autocomplete request performance. [#68](https://github.com/graphene-monitoring/graphene/pull/68)
  * Users have to migrate documents in Elasticsearch with script located in scripts/v1.9.1/es-index-migration.py

# 1.9.0

## Feature
- support Prometheus Lexer. [#62](https://github.com/graphene-monitoring/graphene/pull/62)
- Support basic auth for Elasticsearch client [#64](https://github.com/graphene-monitoring/graphene/pull/64)
- Drop metric if metric id is greater than 512 bytes. It is es key hard limit. [#64](https://github.com/graphene-monitoring/graphene/pull/67)

## Improvement
- improve local key cache [#60](https://github.com/graphene-monitoring/graphene/pull/60)
- include all option on tag value whenever result set is not empty [#59](https://github.com/graphene-monitoring/graphene/pull/59)
- configurable resolution(rollup) [#63](https://github.com/graphene-monitoring/graphene/pull/63)

# 1.8.0

## Feature
- read and write metrics from Cassandra by following offset based scheme [#44](https://github.com/graphene-monitoring/graphene/issues/44)

# 1.7.1

## Feature
- should deserialize converted metric if label value has whitespace [#54](https://github.com/graphene-monitoring/graphene/pull/54)

# 1.7.0

## Feature
- add input kafka type with graphite deserializer [#43](https://github.com/graphene-monitoring/graphene/pull/43)
- search Graphene metrics by tags [#38](https://github.com/graphene-monitoring/graphene/pull/38)
- add prometheus deserializer in kafka input. [#48](https://github.com/graphene-monitoring/graphene/pull/48)

## Improvement
- extract elasticsearch port and protocol to property [#42](https://github.com/graphene-monitoring/graphene/pull/42) Contributed by @jacek-bakowski
- use index property each handler type [#51](https://github.com/graphene-monitoring/graphene/pull/51)

## Bug fix
- increase graphene cassandra handler max queue size [#40](https://github.com/graphene-monitoring/graphene/pull/40)
- fix several bugs [#46](https://github.com/graphene-monitoring/graphene/pull/46)
- fix prometheus deserializer edge case [#51](https://github.com/graphene-monitoring/graphene/pull/51)

# 1.6.0

## Feature
- add InfluxDB write api support [#35](https://github.com/graphene-monitoring/graphene/pull/35)
- forward to another graphene server for zero downtime upgrade [#36](https://github.com/graphene-monitoring/graphene/pull/36)

## Improvement
- add Errors enum for handling exception more easily [#37](https://github.com/graphene-monitoring/graphene/pull/37)

## Bug fix
- fix rotated index with leading zero [#34](https://github.com/graphene-monitoring/graphene/pull/34)

# 1.5.0

## Feature
- add index-based key search handler & query optimizer for performance [#32](https://github.com/graphene-monitoring/graphene/pull/32)

# 1.4.0

## Feature
- add KeyCache interface for checking duplication metric key [#25](https://github.com/graphene-monitoring/graphene/pull/25)
- add helm charts for support easy installation on k8s [#28](https://github.com/graphene-monitoring/graphene/pull/28) Contributed by @hwanjin-jeong
- add nginx docker image for graphene reader's proxy [#27](https://github.com/graphene-monitoring/graphene/pull/27)
- add dynamic template to store each key's parts in keyword type [#30](https://github.com/graphene-monitoring/graphene/pull/30)

## Improvement
- add bulkAsync feature for indexing performance [#20](https://github.com/graphene-monitoring/graphene/pull/20)
- change package name to com.graphene [#31](https://github.com/graphene-monitoring/graphene/pull/31)
- bump up dependency version [#29](https://github.com/graphene-monitoring/graphene/pull/29)

## Bug fix
- ignore unavailable index when read. [#21](https://github.com/graphene-monitoring/graphene/pull/21)
- improve Key Store throughput [#23](https://github.com/graphene-monitoring/graphene/pull/23)
- prevent calling clearScroll if there's no scrollId. [#24](https://github.com/graphene-monitoring/graphene/pull/24)
- fixed some bugs and removed unnecessary codes and operations. [#26](https://github.com/graphene-monitoring/graphene/pull/26)

# 1.3.0

## Feature
- support past metric indexing based on metric timestamp. [#18](https://github.com/graphene-monitoring/graphene/pull/18)
- full support for MapSeries, ReduceSeries functions. [#19](https://github.com/graphene-monitoring/graphene/pull/19)

## Bug fix
- Index is not determined by metric's timestamp when writing. [#17](https://github.com/graphene-monitoring/graphene/issues/17)

# 1.2.1

## Refactoring
- Modify bulk interval unit to milliseconds
- Bump up elasticsearch client version to 6.8.1
- Add auto_expand_replicas property

# 1.2.0

## Feature
- Reduce unnecessary copy operation when using StringBuilder to make response. [#5](https://github.com/graphene-monitoring/graphene/pull/5)
- Support multiple handlers using property. [#6](https://github.com/graphene-monitoring/graphene/pull/6)
- Add the index based key store handler for performance. [#7](https://github.com/graphene-monitoring/graphene/pull/7)
- The Controller just uses the content set in FullHttpResponse and change it to String AGAIN to make response body to be handled by Spring's RestController, which are all unnecessary. [#10](https://github.com/graphene-monitoring/graphene/pull/10)
- Support elasticsearch 6.x.x with index template and create automatically index. [#11](https://github.com/graphene-monitoring/graphene/pull/11)
- Add index rotating feature with the tenant ( daily, weekly ). [#13](https://github.com/graphene-monitoring/graphene/pull/13)
- Support rotating index query. [#14](https://github.com/graphene-monitoring/graphene/pull/14)
- Add multi-key store handler. [#15](https://github.com/graphene-monitoring/graphene/pull/15)

## Bug fix
- If time-series data is empty, Graphene returns ']' single string. [#8](https://github.com/graphene-monitoring/graphene/pull/8)

## Refactoring
- Try entire code refactoring to apply index based key store handler

graphene-docker/infra/cassandra/3.11.5/conf/cassandra-env.sh
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

calculate_heap_sizes()
{
    case "`uname`" in
        Linux)
            system_memory_in_mb=`free -m | awk '/:/ {print $2;exit}'`
            system_cpu_cores=`egrep -c 'processor([[:space:]]+):.*' /proc/cpuinfo`
        ;;
        FreeBSD)
            system_memory_in_bytes=`sysctl hw.physmem | awk '{print $2}'`
            system_memory_in_mb=`expr $system_memory_in_bytes / 1024 / 1024`
            system_cpu_cores=`sysctl hw.ncpu | awk '{print $2}'`
        ;;
        SunOS)
            system_memory_in_mb=`prtconf | awk '/Memory size:/ {print $3}'`
            system_cpu_cores=`psrinfo | wc -l`
        ;;
        Darwin)
            system_memory_in_bytes=`sysctl hw.memsize | awk '{print $2}'`
            system_memory_in_mb=`expr $system_memory_in_bytes / 1024 / 1024`
            system_cpu_cores=`sysctl hw.ncpu | awk '{print $2}'`
        ;;
        *)
            # assume reasonable defaults for e.g. a modern desktop or
            # cheap server
            system_memory_in_mb="2048"
            system_cpu_cores="2"
        ;;
    esac

    # some systems like the raspberry pi don't report cores, use at least 1
    if [ "$system_cpu_cores" -lt "1" ]
    then
        system_cpu_cores="1"
    fi

    # set max heap size based on the following
    # max(min(1/2 ram, 1024MB), min(1/4 ram, 8GB))
    # calculate 1/2 ram and cap to 1024MB
    # calculate 1/4 ram and cap to 8192MB
    # pick the max
    half_system_memory_in_mb=`expr $system_memory_in_mb / 2`
    quarter_system_memory_in_mb=`expr $half_system_memory_in_mb / 2`
    if [ "$half_system_memory_in_mb" -gt "1024" ]
    then
        half_system_memory_in_mb="1024"
    fi
    if [ "$quarter_system_memory_in_mb" -gt "8192" ]
    then
        quarter_system_memory_in_mb="8192"
    fi
    if [ "$half_system_memory_in_mb" -gt "$quarter_system_memory_in_mb" ]
    then
        max_heap_size_in_mb="$half_system_memory_in_mb"
    else
        max_heap_size_in_mb="$quarter_system_memory_in_mb"
    fi
    MAX_HEAP_SIZE="${max_heap_size_in_mb}M"

    # Young gen: min(max_sensible_per_modern_cpu_core * num_cores, 1/4 * heap size)
    max_sensible_yg_per_core_in_mb="100"
    max_sensible_yg_in_mb=`expr $max_sensible_yg_per_core_in_mb "*" $system_cpu_cores`

    desired_yg_in_mb=`expr $max_heap_size_in_mb / 4`

    if [ "$desired_yg_in_mb" -gt "$max_sensible_yg_in_mb" ]
    then
        HEAP_NEWSIZE="${max_sensible_yg_in_mb}M"
    else
        HEAP_NEWSIZE="${desired_yg_in_mb}M"
    fi
}

# Determine the sort of JVM we'll be running on.
java_ver_output=`"${JAVA:-java}" -version 2>&1`
jvmver=`echo "$java_ver_output" | grep '[openjdk|java] version' | awk -F'"' 'NR==1 {print $2}' | cut -d\- -f1`
JVM_VERSION=${jvmver%_*}
JVM_PATCH_VERSION=25

if [ "$JVM_VERSION" \< "1.8" ] ; then
    echo "Cassandra 3.0 and later require Java 8u40 or later."
    exit 1;
fi

if [ "$JVM_VERSION" \< "1.8" ] && [ "$JVM_PATCH_VERSION" -lt 40 ] ; then
    echo "Cassandra 3.0 and later require Java 8u40 or later."
    exit 1;
fi

jvm=`echo "$java_ver_output" | grep -A 1 '[openjdk|java] version' | awk 'NR==2 {print $1}'`
case "$jvm" in
    OpenJDK)
        JVM_VENDOR=OpenJDK
        # this will be "64-Bit" or "32-Bit"
        JVM_ARCH=`echo "$java_ver_output" | awk 'NR==3 {print $2}'`
        ;;
    "Java(TM)")
        JVM_VENDOR=Oracle
        # this will be "64-Bit" or "32-Bit"
        JVM_ARCH=`echo "$java_ver_output" | awk 'NR==3 {print $3}'`
        ;;
    *)
        # Help fill in other JVM values
        JVM_VENDOR=other
        JVM_ARCH=unknown
        ;;
esac

#GC log path has to be defined here because it needs to access CASSANDRA_HOME
JVM_OPTS="$JVM_OPTS -Xloggc:/var/log/cassandra/gc_log/gc.log"

# Here we create the arguments that will get passed to the jvm when
# starting cassandra.

# Read user-defined JVM options from jvm.options file
JVM_OPTS_FILE=$CASSANDRA_CONF/jvm.options
for opt in `grep "^-" $JVM_OPTS_FILE`
do
  JVM_OPTS="$JVM_OPTS $opt"
done

# Check what parameters were defined on jvm.options file to avoid conflicts
echo $JVM_OPTS | grep -q Xmn
DEFINED_XMN=$?
echo $JVM_OPTS | grep -q Xmx
DEFINED_XMX=$?
echo $JVM_OPTS | grep -q Xms
DEFINED_XMS=$?
echo $JVM_OPTS | grep -q UseConcMarkSweepGC
USING_CMS=$?
echo $JVM_OPTS | grep -q UseG1GC
USING_G1=$?

# Override these to set the amount of memory to allocate to the JVM at
# start-up. For production use you may wish to adjust this for your
# environment. MAX_HEAP_SIZE is the total amount of memory dedicated
# to the Java heap. HEAP_NEWSIZE refers to the size of the young
# generation. Both MAX_HEAP_SIZE and HEAP_NEWSIZE should be either set
# or not (if you set one, set the other).
#
# The main trade-off for the young generation is that the larger it
# is, the longer GC pause times will be. The shorter it is, the more
# expensive GC will be (usually).
#
# The example HEAP_NEWSIZE assumes a modern 8-core+ machine for decent pause
# times. If in doubt, and if you do not particularly want to tweak, go with
# 100 MB per physical CPU core.

#MAX_HEAP_SIZE="4G"
#HEAP_NEWSIZE="800M"

# Set this to control the amount of arenas per-thread in glibc
#export MALLOC_ARENA_MAX=4

# only calculate the size if it's not set manually
if [ "x$MAX_HEAP_SIZE" = "x" ] && [ "x$HEAP_NEWSIZE" = "x" -o $USING_G1 -eq 0 ]; then
    calculate_heap_sizes
elif [ "x$MAX_HEAP_SIZE" = "x" ] ||  [ "x$HEAP_NEWSIZE" = "x" -a $USING_G1 -ne 0 ]; then
    echo "please set or unset MAX_HEAP_SIZE and HEAP_NEWSIZE in pairs when using CMS GC (see cassandra-env.sh)"
    exit 1
fi

if [ "x$MALLOC_ARENA_MAX" = "x" ] ; then
    export MALLOC_ARENA_MAX=4
fi

# We only set -Xms and -Xmx if they were not defined on jvm.options file
# If defined, both Xmx and Xms should be defined together.
if [ $DEFINED_XMX -ne 0 ] && [ $DEFINED_XMS -ne 0 ]; then
     JVM_OPTS="$JVM_OPTS -Xms${MAX_HEAP_SIZE}"
     JVM_OPTS="$JVM_OPTS -Xmx${MAX_HEAP_SIZE}"
elif [ $DEFINED_XMX -ne 0 ] || [ $DEFINED_XMS -ne 0 ]; then
     echo "Please set or unset -Xmx and -Xms flags in pairs on jvm.options file."
     exit 1
fi

# We only set -Xmn flag if it was not defined in jvm.options file
# and if the CMS GC is being used
# If defined, both Xmn and Xmx should be defined together.
if [ $DEFINED_XMN -eq 0 ] && [ $DEFINED_XMX -ne 0 ]; then
    echo "Please set or unset -Xmx and -Xmn flags in pairs on jvm.options file."
    exit 1
elif [ $DEFINED_XMN -ne 0 ] && [ $USING_CMS -eq 0 ]; then
    JVM_OPTS="$JVM_OPTS -Xmn${HEAP_NEWSIZE}"
fi

if [ "$JVM_ARCH" = "64-Bit" ] && [ $USING_CMS -eq 0 ]; then
    JVM_OPTS="$JVM_OPTS -XX:+UseCondCardMark"
fi

# provides hints to the JIT compiler
JVM_OPTS="$JVM_OPTS -XX:CompileCommandFile=$CASSANDRA_CONF/hotspot_compiler"

# add the jamm javaagent
JVM_OPTS="$JVM_OPTS -javaagent:$CASSANDRA_HOME/lib/jamm-0.3.0.jar"

# set jvm HeapDumpPath with CASSANDRA_HEAPDUMP_DIR
if [ "x$CASSANDRA_HEAPDUMP_DIR" != "x" ]; then
    JVM_OPTS="$JVM_OPTS -XX:HeapDumpPath=$CASSANDRA_HEAPDUMP_DIR/cassandra-`date +%s`-pid$$.hprof"
fi

# stop the jvm on OutOfMemoryError as it can result in some data corruption
# uncomment the preferred option
# ExitOnOutOfMemoryError and CrashOnOutOfMemoryError require a JRE greater or equals to 1.7 update 101 or 1.8 update 92
# For OnOutOfMemoryError we cannot use the JVM_OPTS variables because bash commands split words
# on white spaces without taking quotes into account
# JVM_OPTS="$JVM_OPTS -XX:+ExitOnOutOfMemoryError"
# JVM_OPTS="$JVM_OPTS -XX:+CrashOnOutOfMemoryError"
JVM_ON_OUT_OF_MEMORY_ERROR_OPT="-XX:OnOutOfMemoryError=kill -9 %p"

# print an heap histogram on OutOfMemoryError
# JVM_OPTS="$JVM_OPTS -Dcassandra.printHeapHistogramOnOutOfMemoryError=true"

# jmx: metrics and administration interface
#
# add this if you're having trouble connecting:
# JVM_OPTS="$JVM_OPTS -Djava.rmi.server.hostname=<public name>"
#
# see
# https://blogs.oracle.com/jmxetc/entry/troubleshooting_connection_problems_in_jconsole
# for more on configuring JMX through firewalls, etc. (Short version:
# get it working with no firewall first.)
#
# Cassandra ships with JMX accessible *only* from localhost.
# To enable remote JMX connections, uncomment lines below
# with authentication and/or ssl enabled. See https://wiki.apache.org/cassandra/JmxSecurity
#
#if [ "x$LOCAL_JMX" = "x" ]; then
#    LOCAL_JMX=yes
#fi

# Specifies the default port over which Cassandra will be available for
# JMX connections.
# For security reasons, you should not expose this port to the internet.  Firewall it if needed.
JMX_PORT="7199"
#JVM_OPTS="$JVM_OPTS -Dcassandra.jmx.local.port=$JMX_PORT"
JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.authenticate=false"
JVM_OPTS="$JVM_OPTS -Dcassandra.jmx.remote.port=$JMX_PORT"
# if ssl is enabled the same port cannot be used for both jmx and rmi so either
# pick another value for this property or comment out to use a random port (though see CASSANDRA-7087 for origins)
JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.rmi.port=$JMX_PORT"

# turn on JMX authentication. See below for further options
JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.authenticate=false"

# jmx ssl options
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.ssl=true"
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.ssl.need.client.auth=true"
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.ssl.enabled.protocols=<enabled-protocols>"
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.ssl.enabled.cipher.suites=<enabled-cipher-suites>"
#JVM_OPTS="$JVM_OPTS -Djavax.net.ssl.keyStore=/path/to/keystore"
#JVM_OPTS="$JVM_OPTS -Djavax.net.ssl.keyStorePassword=<keystore-password>"
#JVM_OPTS="$JVM_OPTS -Djavax.net.ssl.trustStore=/path/to/truststore"
#JVM_OPTS="$JVM_OPTS -Djavax.net.ssl.trustStorePassword=<truststore-password>"

# jmx authentication and authorization options. By default, auth is only
# activated for remote connections but they can also be enabled for local only JMX
## Basic file based authn & authz
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.password.file=/etc/cassandra/jmxremote.password"
#JVM_OPTS="$JVM_OPTS -Dcom.sun.management.jmxremote.access.file=/etc/cassandra/jmxremote.access"
## Custom auth settings which can be used as alternatives to JMX's out of the box auth utilities.
## JAAS login modules can be used for authentication by uncommenting these two properties.
## Cassandra ships with a LoginModule implementation - org.apache.cassandra.auth.CassandraLoginModule -
## which delegates to the IAuthenticator configured in cassandra.yaml.tmpl. See the sample JAAS configuration
## file cassandra-jaas.config
#JVM_OPTS="$JVM_OPTS -Dcassandra.jmx.remote.login.config=CassandraLogin"
#JVM_OPTS="$JVM_OPTS -Djava.security.auth.login.config=$CASSANDRA_CONF/cassandra-jaas.config"

## Cassandra also ships with a helper for delegating JMX authz calls to the configured IAuthorizer,
## uncomment this to use it. Requires one of the two authentication options to be enabled
#JVM_OPTS="$JVM_OPTS -Dcassandra.jmx.authorizer=org.apache.cassandra.auth.jmx.AuthorizationProxy"

# To use mx4j, an HTML interface for JMX, add mx4j-tools.jar to the lib/
# directory.
# See http://cassandra.apache.org/doc/3.11/operating/metrics.html#jmx
# By default mx4j listens on 0.0.0.0:8081. Uncomment the following lines
# to control its listen address and port.
#MX4J_ADDRESS="-Dmx4jaddress=127.0.0.1"
#MX4J_PORT="-Dmx4jport=8081"

# Cassandra uses SIGAR to capture OS metrics CASSANDRA-7838
# for SIGAR we have to set the java.library.path
# to the location of the native libraries.
JVM_OPTS="$JVM_OPTS -Djava.library.path=$CASSANDRA_HOME/lib/sigar-bin"

JVM_OPTS="$JVM_OPTS $MX4J_ADDRESS"
JVM_OPTS="$JVM_OPTS $MX4J_PORT"
JVM_OPTS="$JVM_OPTS $JVM_EXTRA_OPTS"

JVM_OPTS="$JVM_OPTS -Dcassandra.max_queued_native_transport_requests=8192"
